### PR TITLE
:bug: Source Stripe: do not emit empty state messages

### PR DESCRIPTION
**Pull request by davydov-d**
## What
Emit non-empty state messages
https://github.com/airbytehq/oncall/issues/3004

## How
- Bugfix: do not return empty state when that same empty state was passed in. Instead, return the cursor value from the latest record
- Another bugfix for streams that did not support incremental syncs until version 4.0.0 and do not have a cursor value in a full refresh or initial incremental sync mode: fill the cursor value with current timestamp
- Added unit tests
- Minor doc update
- Updated stream schema + expected records for the build to pass

## 🚨 User Impact 🚨
No impact

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-stripe test report (commit [`58e4fe4606`](https://github.com/airbytehq/airbyte/pull/30660/commits/58e4fe46064db1132b5727342ed6bb9eff370522)) - ❌

⏲️  Total pipeline duration: 11mn57s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-stripe docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ❌        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-stripe/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/ddavydov_oc-3004-source-stripe-fix-incremental-sync-state/1695303637/58e4fe46064db1132b5727342ed6bb9eff370522/source-stripe/4.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:58e4fe46064db1132b5727342ed6bb9eff370522)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-stripe test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-stripe test report (commit [`c65e13fdfa`](https://github.com/airbytehq/airbyte/pull/30660/commits/c65e13fdfa63fb25973d9ee939dc9301b201acfb)) - ✅

⏲️  Total pipeline duration: 11mn57s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-stripe docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ✅        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-stripe/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/ddavydov_oc-3004-source-stripe-fix-incremental-sync-state/1695307985/c65e13fdfa63fb25973d9ee939dc9301b201acfb/source-stripe/4.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:c65e13fdfa63fb25973d9ee939dc9301b201acfb)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-stripe test
```

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py [hunk 1]
```diff
@@ -55,8 +55,16 @@ def __init__(self, cursor_field: str, legacy_cursor_field: Optional[str]):
...
+                yield record
+                continue
+            if self.legacy_cursor_field:
                 record[self.cursor_field] = record[self.legacy_cursor_field]
+                yield record
+                continue
+            # this is done for the initial incremental sync to emit valid state message as
+            # some streams do not have cursor values when running initial incremental or full refresh sync.
+            record[self.cursor_field] = pendulum.now().int_timestamp
             yield record
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:68]:
```suggestion
            if self.cursor_field in record:
                yield record
                continue  # Skip the rest of the loop iteration

            # fetch legacy_cursor_field from record; default to current timestamp for initial syncs without an any cursor.
            current_cursor_value = record.get(self.legacy_cursor_field, pendulum.now().int_timestamp)
        
            # yield the record with the added cursor_field
            yield record | {self.cursor_field: current_cursor_value} 
```

**Review by lazebnyi — APPROVED** [review]
LGTM

**Comment by airbyte-oss-build-runner** [comment]
##  source-stripe test report (commit [`cb9039b07a`](https://github.com/airbytehq/airbyte/pull/30660/commits/cb9039b07a1a0cf1bc80f8d3785b4a6d80e589ef)) - ❌

⏲️  Total pipeline duration: 11mn52s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-stripe docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ✅        |
| Code format checks                                                   | ❌        |
| Validate airbyte-integrations/connectors/source-stripe/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/ddavydov_oc-3004-source-stripe-fix-incremental-sync-state/1695314046/cb9039b07a1a0cf1bc80f8d3785b4a6d80e589ef/source-stripe/4.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:cb9039b07a1a0cf1bc80f8d3785b4a6d80e589ef)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-stripe test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-stripe test report (commit [`57c3dcd5f3`](https://github.com/airbytehq/airbyte/pull/30660/commits/57c3dcd5f32bf74a49f198a3127fc4ffe6197a86)) - ✅

⏲️  Total pipeline duration: 12mn08s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-stripe docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ✅        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-stripe/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/ddavydov_oc-3004-source-stripe-fix-incremental-sync-state/1695323232/57c3dcd5f32bf74a49f198a3127fc4ffe6197a86/source-stripe/4.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:57c3dcd5f32bf74a49f198a3127fc4ffe6197a86)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-stripe test
```

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py [hunk 2]
```diff
@@ -55,9 +55,15 @@ def __init__(self, cursor_field: str, legacy_cursor_field: Optional[str]):
...
-            yield record
+            if self.cursor_field in record:
+                yield record
+                continue  # Skip the rest of the loop iteration
+
+            # fetch legacy_cursor_field from record; default to current timestamp for initial syncs without an any cursor.
+            current_cursor_value = record.get(self.legacy_cursor_field, pendulum.now().int_timestamp)
+
+            # yield the record with the added cursor_field
+            yield record | {self.cursor_field: current_cursor_value}
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:66]: Defaulting on `now` seems odd to me. Is it only some records in a stream that might not have `self.cursor_field` and `self.legacy_cursor_field` in the record? I would be afraid of the following series of events:
* Consume a record #&NoBreak;1 `self.cursor_field` and `self.legacy_cursor_field` not in the record so we use `now`
* checkpointing from AbstractSource
* Consume a record #&NoBreak;2 with `self.cursor_field` or `self.legacy_cursor_field` less than `now`
* Sync crash

In that case, we would start the next sync at the `now` when we consumed record #&NoBreak;1 but would potentially miss record #&NoBreak;2 and others. The only reason allow for `now` as the default is if all the records for the same stream will not have `self.legacy_cursor_field` as a field. Can you confirm?

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py [hunk 3]
```diff
@@ -55,9 +55,15 @@ def __init__(self, cursor_field: str, legacy_cursor_field: Optional[str]):
     def extract_records(self, response: requests.Response) -> Iterable[MutableMapping]:
         records = super().extract_records(response)
         for record in records:
-            if self.legacy_cursor_field and self.cursor_field not in record:
-                record[self.cursor_field] = record[self.legacy_cursor_field]
-            yield record
+            if self.cursor_field in record:
+                yield record
+                continue  # Skip the rest of the loop iteration
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:60]: nit: Would it be more readable to use the `else` statement instead of having to add a comment?

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py [hunk 4]
```diff
@@ -443,7 +449,7 @@ def parent_stream(self, stream):
 
     @property
     def cursor_field(self) -> Union[str, List[str]]:
-        return [self._cursor_field]
+        return self._cursor_field
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:452]: Why this change?

File: airbyte-integrations/connectors/source-stripe/unit_tests/test_streams.py [hunk 5]
```diff
@@ -549,3 +549,18 @@ def test_cursorless_incremental_substream(requests_mock, bank_accounts, sync_mod
     for slice_ in stream.stream_slices(sync_mode=sync_mode, stream_state=stream_state):
         for record in stream.read_records(sync_mode=sync_mode, stream_state=stream_state, stream_slice=slice_):
             stream.get_updated_state(stream_state, record)
+
+
+@pytest.mark.parametrize("stream", ("bank_accounts",))
+def test_get_updated_state(stream, request, requests_mock):
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-stripe/unit_tests/test_streams.py:555]: Should we have a test that validates that if a record does not have `cursor_field` and `legacy_cursor_field`, we use `now` as a `cursor_field`?

**Review by maxi297 — COMMENTED** [review]
There is a couple of things I don't understand enough to give approval. With the answer to my question, I should be able to approve

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py (hunk 2 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:66]: @maxi297 yes, the case here is *some* streams that do not have a cursor value at all when running in a full refresh mode (or initial incremental sync), so the consequences you described are not relevant

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py (hunk 4 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:452]: Because we're using `current_state[self.cursor_field]` in the code below

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py (hunk 3 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:60]: matter of taste :) I prefer early exit/continue over else statements

Comment (davydov-d) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:60]: but if you insist, I can change it of course

**Comment by davydov-d** [comment]
@maxi297 can you please take a look at the comments please

**Review by maxi297 — APPROVED** [review]
Given the added information by @davydov-d, I have enough information to approve this change

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py (hunk 3 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:60]: There are a couple of resources regarding cognitive complexity recommending to avoid "jumps to" labels. One example is [Sonar](https://www.sonarsource.com/docs/CognitiveComplexity.pdf?_gl=1*rr0zbd*_gcl_au*NDY5NzI2ODIxLjE2OTU0MDUxMDQ.) I personally agree with these but I'll not enforce it if the maintainers of the code prefer otherwise

Comment (davydov-d) [reply airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:60]: Thank you! I will leave it here as is, but take into account in the future updates
